sdk/core/Azure.Core.Experimental/src/JsonData.cs

```diff
@@ -260,315 +269,17 @@ public string ToJsonString()
-            EnsureArray().Add(value);
-            return value;
-        }
-
-        /// <summary>
-        /// Inserts a new value at the end of an array.
-        /// </summary>
-        /// <param name="serializable">The value to insert into the array.</param>
-        /// <returns>A <see cref="JsonData"/> of the serialized object.</returns>
-        /// <remarks>
-        /// If the <see cref="Kind"/> property is not <see cref="JsonValueKind.Array"/> this method throws <see cref="InvalidOperationException"/>.
-        /// </remarks>
-        public JsonData Add<T>(T[] serializable)
-        {
-            JsonData value = new JsonData(serializable.Select(x => new JsonData(x)));
-            EnsureArray().Add(value);
-            return value;
-        }
-
-        /// <summary>
-        /// Inserts a new value at the end of an array.
-        /// </summary>
-        /// <param name="serializable">The value to insert into the array.</param>
-        /// <param name="options">Options to control the conversion behavior.</param>
-        /// <returns>A <see cref="JsonData"/> of the serialized object.</returns>
-        /// <remarks>
-        /// If the <see cref="Kind"/> property is not <see cref="JsonValueKind.Array"/> this method throws <see cref="InvalidOperationException"/>.
-        /// </remarks>
-        public JsonData Add<T>(T[] serializable, JsonSerializerOptions options)
-        {
-            JsonData value = new JsonData(serializable.Select(x => new JsonData(x, options)));
-            EnsureArray().Add(value);
-            return value;
-        }
-
-        /// <summary>
-        /// Inserts a new empty object at the end of an array.
-        /// </summary>
-        /// <returns>A <see cref="JsonData"/> for the newly created empty object.</returns>
-        /// <remarks>
-        /// If the <see cref="Kind"/> property is not <see cref="JsonValueKind.Array"/> this method throws <see cref="InvalidOperationException"/>.
-        /// </remarks>
-        public JsonData AddEmptyObject()
-        {
-            JsonData value = EmptyObject();
-            EnsureArray().Add(value);
-            return value;
-        }
-
-        /// <summary>
-        /// Inserts a new empty array at the end of an array.
-        /// </summary>
-        /// <returns>A <see cref="JsonData"/> for the newly created empty array.</returns>
-        /// <remarks>
-        /// If the <see cref="Kind"/> property is not <see cref="JsonValueKind.Array"/> this method throws <see cref="InvalidOperationException"/>.
-        /// </remarks>
-        public JsonData AddEmptyArray()
-        {
-            JsonData value = EmptyArray();
-            EnsureArray().Add(value);
-            return value;
+            using var memoryStream = new MemoryStream();
+            using (var writer = new Utf8JsonWriter(memoryStream))
+            {
+                WriteTo(writer);
+            }
+            return new BinaryData(memoryStream.ToArray());
```

Comment:
How about?

```suggestion
            return new BinaryData(memoryStream.GetBuffer().AsMemory(0, (int)stream.Position));
```

Reply:
This is how the BinaryData FromStream method constructs the instance after copying into a separate MemoryStream - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.Memory.Data/src/BinaryData.cs#L179

I think the reason we didn't type sniff for MemoryStream and attempt to extract the ArraySegment is that we wanted the semantics to be consistent across different streams.